cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -662,6 +663,84 @@ public void checkCustomCaSecret(CertificateAuthority ca, Secret certSecret, Secr
+            String zkSsName = ZookeeperCluster.zookeeperClusterName(name);
+
+            Future<StatefulSet> kafkaSsFuture = kafkaSetOperations.getAsync(namespace, kafkaSsName);
+            Future<StatefulSet> zkSsFuture = zkSetOperations.getAsync(namespace, zkSsName);
+
+            return CompositeFuture.all(kafkaSsFuture, zkSsFuture).compose(allFutures -> {
+                if (allFutures.succeeded()) {
+
+                    StatefulSet zkSs = zkSsFuture.result();
+                    StatefulSet kafkaSs = kafkaSsFuture.result();
+
+                    if (zkSs == null || kafkaSs == null) {
+                        return Future.succeededFuture(this);
+                    }
+
+                    KafkaVersionChange versionChange = getKafkaVersionChange(kafkaSs);
+
+                    log.debug("Checking if we need to change the Zookeeper image for Zookeeper cluster {}", zkSsName);
+
+                    if (versionChange.isNoop()) {
+                        log.debug("Kafka.spec.kafka.version is unchanged therefore no change to Zookeeper is required");
+                        return Future.succeededFuture(this);
+                    } else {
+                        if (versionChange.isDowngrade()) {
+                            log.debug("Kafka downgrade from {} to {} requires Zookeeper downgrade from {} to {}",
+                                    versionChange.from().version(),
+                                    versionChange.to().version(),
+                                    versionChange.from().zookeeperVersion(),
+                                    versionChange.to().zookeeperVersion());
+                        } else {
+                            log.debug("Kafka upgrade from {} to {} requires Zookeeper upgrade from {} to {}",
+                                    versionChange.from().version(),
+                                    versionChange.to().version(),
+                                    versionChange.from().zookeeperVersion(),
+                                    versionChange.to().zookeeperVersion());
+                        }
+                        return updateZkStatefulSet(zkSs, kafkaAssembly, versionChange).map(this);
+                    }
+                } else {
+                    return Future.failedFuture("No kafka and/or ZK StatefulSet information was obtainable");
+                }
+            });
+        }
+
+        private Future<Void> updateZkStatefulSet(StatefulSet zookeeperSs, Kafka kafkaAssembly, KafkaVersionChange versionChange) {
+
+            return waitForQuiescence(zookeeperSs).compose(v -> {
+                // Get the zookeeper image currently set in the Kafka CR or, if that is not set, the image from the target Kafka version
+                String newZkImage = versions.kafkaImage(kafkaAssembly.getSpec().getZookeeper().getImage(), versionChange.to().version());
+
+                // TODO: Do we need to pull in any config changes for the new version of zk?
```

Comment:
Is this still a TODO? If yes, either open an issue for it or implement it. Or should be removed?

Reply:
No this is no longer relevant, this is handled in the second ZK roll so I will remove this.